File: SDKs/PopcornNativeSDK/PopcornCore/Domain/Commons/OptionalWrapper.swift
//
//  OptionalWrapper.swift
//  Popcorn
//
//  Created by Eduardo Raupp Peretto on 13/05/25.
//

import Foundation

public enum OptionalWrapper<T: SelectionPickable>: Hashable {
    case none
    case some(T)

    public var description: String {
        switch self {
        case .none: "Not selected"
        case let .some(value): value.name
        }
    }

    public init(_ value: T?) {
        if let value {
            self = .some(value)
        } else {
            self = .none
        }
    }
}

extension OptionalWrapper: CaseIterable {
    public static var allCases: [OptionalWrapper] {
        var allCases = T.allCases.map { self.some($0) }
        allCases.append(.none)
        return allCases
    }
}

File: KotlinShared/shared/src/commonMain/kotlin/com/eduardoperetto/popcornshared/domain/OptionalWrapper.kt
package com.eduardoperetto.popcornshared.domain

// OptionalWrapper equivalent
sealed class OptionalWrapper<T : SelectionPickable> {
    data class Some<T : SelectionPickable>(val value: T) : OptionalWrapper<T>()
    class None<T : SelectionPickable> : OptionalWrapper<T>()

    val description: String
        get() = when (this) {
            is None -> "Not selected"
            is Some -> value.name
        }

    companion object {
        fun <T : SelectionPickable> from(value: T?): OptionalWrapper<T> =
            if (value != null) Some(value) else None()

        fun <T : SelectionPickable> allCases(all: List<T>): List<OptionalWrapper<T>> {
            val cases = all.map { Some(it) }
            return cases + None()
        }
    }
}